server/src/main/kotlin/org/itmodreamteam/myrest/server/service/reservation/ReservationServiceImpl.kt

```diff
@@ -0,0 +1,100 @@
+import org.itmodreamteam.myrest.server.model.restaurant.Reservation
+import org.itmodreamteam.myrest.server.model.restaurant.Restaurant
+import org.itmodreamteam.myrest.server.model.restaurant.RestaurantTable
+import org.itmodreamteam.myrest.server.repository.restaurant.EmployeeRepository
+import org.itmodreamteam.myrest.server.repository.restaurant.ReservationRepository
+import org.itmodreamteam.myrest.server.security.CurrentUserService
+import org.itmodreamteam.myrest.server.service.notification.NotificationService
+import org.itmodreamteam.myrest.shared.restaurant.ReservationStatus
+import org.springframework.stereotype.Service
+import java.time.LocalDateTime
+
+@Service
+class ReservationServiceImpl(
+    private val reservationRepository: ReservationRepository,
+    private val notificationService: NotificationService,
+    private val currentUserService: CurrentUserService,
+    private val employeeRepository: EmployeeRepository
+) : ReservationService {
+
+    override fun submitReservationForApproval(
+        table: RestaurantTable,
+        activeFrom: LocalDateTime,
+        activeUntil: LocalDateTime
+    ): Reservation {
+        val overlapping = findOverlappingReservations(table, activeFrom, activeUntil)
+        if (overlapping.isNotEmpty()) {
+            throw UserException("Временной интервал $activeFrom - $activeUntil занят. Попробуйте другое время")
+        }
+        var reservation = Reservation(currentUserService.currentUserEntity, table, activeFrom, activeUntil)
+        reservation = reservationRepository.save(reservation)
+        /*
+        TODO
+        It's possible to have concurrent save operation which can lead to overlapping reservations (due to absence of something like time range overlapping constraint).
+        Due to caching on application side it's not possible to guarantee consistency.
+        Possible solutions are:
+        - Trigger check on DB side
+        - Batch job detecting overlapping reservations and notifying responsible managers
+        - Nested transaction + check with possible fallback (need to setup transaction manager)
+        */
+        return reservation
+    }
+
+    override fun findOverlappingReservations(
+        table: RestaurantTable,
+        activeFrom: LocalDateTime,
+        activeUntil: LocalDateTime
+    ): List<Reservation> {
+        return reservationRepository.findReservationsForTableByStatusesAndTimeRangeOverlapping(
+            table,
+            listOf(ReservationStatus.PENDING, ReservationStatus.APPROVED, ReservationStatus.IN_PROGRESS),
+            activeFrom, activeUntil
+        )
+    }
+
+    override fun reject(reservation: Reservation): Reservation {
+        return updateStatusAndPersist(reservation, ReservationStatus.REJECTED)
+    }
+
+    override fun approve(reservation: Reservation): Reservation {
+        reservation.manager = getCurrentUserBeingManagerOf(reservation.table.restaurant)
```

Comment:
The fact of current user being manager of this restaurant is already verified in ReservationPermissionEvaluator